chore: migrate to eslint

klaascuvelier opened the pull request:
Closes carbon-design-system/carbon-components-angular#3065

This PR replaces tslint with eslint. Config stays as close to current config is possible.

Akshat55 commented on src/notification/actionable-notification.component.ts:
```diff
@@ -92,10 +92,14 @@ export class ActionableNotification extends BaseNotification {
 	@HostBinding("class.cds--actionable-notification--error") get isError() { return this.notificationObj.type === "error"; }
 	@HostBinding("class.cds--actionable-notification--info") get isInfo() { return this.notificationObj.type === "info"; }
-	@HostBinding("class.cds--actionable-notification--info-square") get isInfoSquare() { return this.notificationObj.type === "info-square"; }
+	@HostBinding("class.cds--actionable-notification--info-square") get isInfoSquare() {
+		return this.notificationObj.type === "info-square";
+	}
```
I'm assuming we are using predefined line length since I don't see the rule for it, might be good to override to what it was before? 🤔

Akshat55 commented on src/popover/popover.directive.ts:
```diff
@@ -153,8 +153,12 @@ export class PopoverContainer implements AfterViewInit, OnChanges, OnDestroy {
 					const height = computedStyle.getPropertyValue("--cds-popover-caret-height");
-					this.caretOffset = (offset?.includes("px") ? Number(offset.split("px", 1)[0]) : Number(offset.split("rem", 1)[0]) * 16) || 10;
-					this.caretHeight = (height?.includes("px") ? Number(height.split("px", 1)[0]) : Number(height.split("rem", 1)[0]) * 16) || 6;
+					this.caretOffset = (offset?.includes("px")
+						? Number(offset.split("px", 1)[0])
+						: Number(offset.split("rem", 1)[0]) * 16) || 10;
+					this.caretHeight = (height?.includes("px")
+						? Number(height.split("px", 1)[0])
+						: Number(height.split("rem", 1)[0]) * 16) || 6;
```
Seems harder to read now? Could just be how it's coded that makes it harder to read?

Thoughts on this?

klaascuvelier replied:
<img width="736" alt="Screenshot 2025-02-26 at 14 33 13" src="https://github.com/user-attachments/assets/76db0c7e-54c6-418a-8c3e-cce062fbc389" />

This is because of the line length too. Another case that was not caught in tslint. 
I am used to seeing this structure because of Prettier in my other projects, so for me this is actually more clear. 

We could slightly increase the line length to resolve a couple of changes I did?

Akshat55 commented on src/popover/popover.stories.ts:
```diff
@@ -140,7 +138,13 @@ const AutoAlignTemplate = (args) => ({
+							xmlns="http://www.w3.org/2000/svg"
+							width="16"
+							height="16"
+							viewBox="0 0 32 32"
+						>
```
```suggestion
							viewBox="0 0 32 32">
```

Akshat55 commented on src/search/search.component.spec.ts:
```diff
@@ -23,7 +23,7 @@ describe("Search", () => {
 			providers: []
 		})
-		.compileComponents();
+			.compileComponents();
```
```suggestion
		.compileComponents();
```

Looks cleaner to keep it aligned?

klaascuvelier replied:
It's possible to allow this, but then I need to turn of the `indent` rule for MemberExpressions. 

This however can lead to situations like this:
<img width="1238" alt="Screenshot 2025-02-26 at 14 40 53" src="https://github.com/user-attachments/assets/55b10376-6583-466c-bb40-e8d20a81cbbc" />

Akshat55 commented on src/tiles/selection-tile.component.ts:
```diff
@@ -11,8 +11,9 @@ import { I18n } from "carbon-components-angular/i18n";
 	selector: "cds-selection-tile, ibm-selection-tile",
-	template: `
-		<input
+	template:
+		/* eslint-disable max-len */
+		`<input
```
Any way we can ignore this for templates?

klaascuvelier replied:
I've tried looking for a way to achieve this but wihtout success so far.
It's probably also not the best idea to remove the length restriction for the templates. I'd suggest to for now just keep it and disable in cases it makes sense?

Akshat55 reviewed (changes requested):
Looks good, just unsure about some of the rules and changes. Do let me know your thoughts on my suggestions.

klaascuvelier commented on src/code-snippet/code-snippet.component.ts:
```diff
@@ -67,7 +67,12 @@ export enum SnippetType {
+						(scroll)="(display === 'multi' ? handleScroll() : null)"
+					>
+						<code #code><ng-container *ngTemplateOutlet="codeTemplate" /></code>
+					</pre>
```
Good catch, this was not eslint, probably my editor

klaascuvelier commented on src/context-menu/context-menu.component.ts:
```diff
@@ -75,7 +75,7 @@ export class ContextMenuComponent implements OnChanges {
 			Array.from(list.querySelectorAll(".cds--menu-item")) as HTMLElement[])
 			.filter(menuItem => !subMenus.some(subMenu => subMenu.contains(menuItem))
-		);
+			);
```
Wrapping bracket are required for the `as HTMLElement[]`, but I formatted it a bit nicer

klaascuvelier commented on src/dialog/dialog.component.ts:
```diff
@@ -137,7 +137,7 @@ export class Dialog implements OnInit, AfterViewInit, OnDestroy {
 					}
 				}
-			);
+				);
```
I think it tries to align the `)` with the start of the `.subscribe`.
I fixed it by placing the `)` behind the `}`.

(This is typically the kind of thing that a formatter like prettier would tackle btw.)

klaascuvelier commented:
Hey @Akshat55 thanks for the feedback. It took me a while to able to get to it, thank you for your patience. 
I fixed issues where it was clear what to do, and added some follow up at other please. Let me know what you think, I'll follow up again asap :)

zvonimirfras commented on src/dialog/overflow-menu/overflow-menu.stories.ts:
```diff
@@ -89,10 +87,18 @@ const LinkTemplate = (args) => ({
+				<cds-overflow-menu-option
+					href="https://www.ibm.com"
+					target="_blank"
+					(selected)="selected($event)"
+				>Option 2</cds-overflow-menu-option>
```
```suggestion
					(selected)="selected($event)">
					Option 2
					</cds-overflow-menu-option>
```

zvonimirfras commented on src/dialog/overflow-menu/overflow-menu.stories.ts:
```diff
@@ -89,10 +87,19 @@ const LinkTemplate = (args) => ({
+				<cds-overflow-menu-option
+					href="https://www.ibm.com"
+					disabled="true"
+					(selected)="selected($event)"
+				>Disabled</cds-overflow-menu-option>
```
```suggestion
					(selected)="selected($event)">
					Disabled
				</cds-overflow-menu-option>
```

zvonimirfras commented on src/dialog/overflow-menu/overflow-menu.stories.ts:
```diff
@@ -89,10 +87,20 @@ const LinkTemplate = (args) => ({
+					href="https://www.ibm.com"
+					target="_blank"
+					(selected)="selected($event)">
+					Option 2
+					</cds-overflow-menu-option>
```
```suggestion
					Option 2
				</cds-overflow-menu-option>
```

klaascuvelier commented:
Thanks for the follow-up, @zvonimirfras and @Akshat55 
I guess this one get be closed now: https://github.com/carbon-design-system/carbon-components-angular/issues/3065
